open_lm/train.py

```diff
@@ -228,11 +236,37 @@ def train_one_epoch(model, data, loss, epoch, step, optimizer, scaler, scheduler
                             * inputs_ii.shape[0]
                             / inputs.shape[0]
                         )
+                    if args.moe_freq > 0:
+                        moe_args = MoEArgs(
+                            hidden_size=model.dim,
+                            ffn_hidden_size=model.dim * 4,
+                            moe_num_experts=args.moe_num_experts,
+                            num_layers=model.n_layers // 2,
+                            moe_expert_model_parallelism=True,
+                            moe_top_k=args.moe_top_k,
+                            device=torch.distributed.get_rank(),
+                            moe_capacity_factor=args.moe_capacity_factor,
+                            moe_loss_weight=args.moe_loss_weight,
+                            fp16=False,
+                        )
+                        local_load_balancing_loss = batched_load_balancing_loss(moe_args)
+                        clear_load_balancing_loss()
+
+                        local_loss = local_load_balancing_loss + local_loss
+
                     backward(local_loss, scaler)
                 if ii == 0:
-                    total_loss = local_loss
+                    if args.moe_freq > 0:
+                        total_loss = local_loss - local_load_balancing_loss
```

Comment:
Why do we subtract load balancing loss?

Reply:
Same question. I guess maybe this is so we can log the loss (without load balancing) separately from the load balancing loss? But if so, let's just subtract this outside of the grad accum loop when logging, instead?

Reply:
yep that is correct @achalddave . I modified this to an "lm_loss" and "load balancing loss" and removed the subtraction so its clearer